src/include/txn.i

```diff
@@ -60,16 +60,51 @@ __wt_txn_context_check(WT_SESSION_IMPL *session, bool requires_txn)
     return (0);
 }
 
+/*
+ * __wt_txn_err_set --
+ *     Set an error in the current transaction.
+ */
+static inline void
+__wt_txn_err_set(WT_SESSION_IMPL *session, int ret)
+{
+    WT_TXN *txn;
+
+    txn = &session->txn;
+
+    /*  Ignore standard errors that don't fail the transaction. */
+    if (ret == WT_NOTFOUND || ret == WT_DUPLICATE_KEY || ret == WT_PREPARE_CONFLICT)
+        return;
+
+    /* Less commonly, it's not a running transaction. */
+    if (!F_ISSET(txn, WT_TXN_RUNNING))
+        return;
+
+    /* The transaction has to be rolled back. */
+    F_SET(txn, WT_TXN_ERROR);
+
+    /*
+     * Check for a prepared transaction, and quit: we can't ignore the error and we can't roll back
+     * a prepared transaction.
+     */
+    if (F_ISSET(txn, WT_TXN_PREPARE))
+        WT_PANIC_MSG(session, ret,
+          "transactional error logged after transaction was prepared, failing the system");
+}
+
 /*
  * __wt_txn_err_chk --
  *     Check the transaction hasn't already failed.
  */
 static inline int
 __wt_txn_err_chk(WT_SESSION_IMPL *session)
 {
-    /* Allow transaction rollback, but nothing else. */
-    if (!F_ISSET(&(session->txn), WT_TXN_ERROR) ||
-      strcmp(session->name, "rollback_transaction") != 0)
+    /*
+     * Allow only a few select operations after an error: WT_CURSOR.{close,reset} are called
+     * internally and WT_SESSION.rollback_transaction is the expected application resolution.
+     */
+    if (!F_ISSET(&(session->txn), WT_TXN_ERROR) || strcmp(session->name, "WT_CURSOR.close") == 0 ||
+      strcmp(session->name, "WT_CURSOR.reset") == 0 ||
+      strcmp(session->name, "WT_SESSION.rollback_transaction") == 0)
```

Comment:
Doing string comparison against `session->name` doesn't feel right to me - we usually use that string for diagnostics, not for making decisions. Can you think of another reasonable solution?

Reply:
This is a little more complicated than [the similar change in the WT-5387 PR](https://github.com/wiredtiger/wiredtiger/pull/5099/commits/d63513c6f2d09911821be7bfb5a8cd62e9c004c3) because it involves not just a `WT_SESSION` method, but also `WT_CURSOR` methods, there's not just a single flag.

I don't think it's a performance problem because we only check if `WT_TXN_ERROR` is already set, which should be rare.

If you prefer adding flags to both structures, or a structure boolean or some other solution, let me know and I'll make the change.

Reply:
You are right - I don't have an obvious better way. The way we've done this before is by splitting macros and using different variants depending on need, but that's pretty ugly and this code is moving away from those API level macros (which I think is generally good).
So I'm happy with it as-is, but will keep looking for ways to avoid using strcmp against the interface names saved in the session handle.